### DISCUSSION TITLE
Horizontal emoji flip?

### DISCUSSION THREAD

**Discussion by ClaireCJS** [microsoft/terminal#16582] [Ideas]
I love that we can italicize emojis -- I love the ANSI & emoji support offered by Windows Terminal.

But it would be really nice to be able to flip them, too.

An example is, to have trumpets on both sides facing towards the message:
![image](https://github.com/microsoft/terminal/assets/789591/8101f5f7-93ec-4fa7-9d11-e582f97578b2)

Currently it's impossible, because the trumpets only point to the right.

Seems like something to extend into the undefined ANSI code space....

**Comment by j4james** [comment]
There's actually already a mechanism in Unicode for handling this sort of thing: [Emoji Glyph Facing Direction](https://www.unicode.org/reports/tr51/tr51-25.html#Direction). 

But this depends on a couple of Unicode features (zero-width-joiners and variant selectors), which aren't yet supported in Windows Terminal. It would probably also require WT to support the proposed Unicode Core mode to get the ZWJ sequence collapsed down to two cells (by default it would be expected to occupy four cells).

**Comment by ClaireCJS** [comment]
wow, very ineresting. i wish it was as easy as an ansi escape sequence.

**Reply by j4james** [reply]
I don't know about easy :) but if it's escape sequences you want, you could always define a custom font like this (bash syntax):
```
printf '\eP0;8;1;10;0;2;20;0{ @???~}{wo_?/[MFJp@@@@@/GCA@??????/??????????;?_ow{}~???/@@@@@pJFM[/??????@ACG/??????????;??????????/??_ogKYr}{/?_ow\\NVb`O/@B@???????;??????????/{}rYKgo_??/O`bVN\\wo_?/???????@B@\e\\'
```

And then your trumpet message could be generated like this:
```
printf '\e( @\e[33m*(\e(B \e[36mThis is an important message \e( @\e[33m)+\e(B\e[m\n'
```

For PowerShell, the syntax would be something like this for the font definition:
```
echo "`eP0;8;1;10;0;2;20;0{ @???~}{wo_?/[MFJp@@@@@/GCA@??????/??????????;?_ow{}~???/@@@@@pJFM[/??????@ACG/??????????;??????????/??_ogKYr}{/?_ow\NVb``O/@B@???????;??????????/{}rYKgo_??/O``bVN\wo_?/???????@B@`e\"
```

And this for the message:
```
echo "`e( @`e[33m*(`e(B `e[36mThis is an important message `e( @`e[33m)+`e(B`e[m"
```

End result:
![image](https://github.com/microsoft/terminal/assets/4181424/7a3fcb23-3c5f-4768-acd2-05d781770ec9)

**Reply by zadjii-msft** [reply]
this is low key one of the more amazing threads on the repo. It's a great practical example of how these sequences might actually be used by someone. Not just docs and specs, but real-world application.

**Reply by ClaireCJS** [reply]
I agree, Zadjii-Msft!  

I just created a pentacle and pentagram character.  They aren't spectacular looking at 10x20 [and i'm wondering if we can give sizes greater than 10x20, maybe James can answer that?] but i'm quite happy to finally have pentagrams after decades of listening to metal music :)

```
set ESCAPE=%@CHAR[27]

echo %ESCAPE%P0;8;1;10;0;2;20;0{ @???~}{wo_?/[MFJp@@@@@/GCA@??????/??????????;?_ow{}~???/@@@@@pJFM[/??????@ACG/??????????;??????????/??_ogKYr}{/?_ow\\NVb`O/@B@???????;??????????/{}rYKgo_??/O`bVN\\wo_?/???????@B@;?_OGCCAaYM/}BEIQayFAA/BKO_oM`PIE/____@@AAAA@B@;YaACCGO_??/AFyaQIEB}?/IP`Mo_OKB?/AAA@@?????@B@;?_OG[cIQaA/}@?_OJ{CAB/BMRaAAANqa/????@@AAAB@B@;aQIc[GO_??/BC{JO_?@}?/qNAAAaRMB?/AAA@@?????%ESCAPE%

set PENTACLE=%ESCAPE( @,-%ESCAPE%(B
set PENTAGRAM=%ESCAPE( @./%ESCAPE%(B
           
```

![image](https://github.com/microsoft/terminal/assets/789591/29958371-8680-4c56-a0b1-4f846fb76cd8)

![image](https://github.com/microsoft/terminal/assets/789591/fc049fe3-f6a6-4e1a-8a4f-4b7e57752910)
![image](https://github.com/microsoft/terminal/assets/789591/66d33df4-5e13-4e6a-91c4-2a6e32359f93)

**Reply by j4james** [reply]
> i'm wondering if we can give sizes greater than 10x20

The max size on Windows Terminal is 16x32, which might be a little better than 10x20. Although be aware that that's not one of the standard sizes, so it may not work on other terminals. If it's just for personal use, though, you probably won't care.

**Comment by ClaireCJS** [comment]
WAIT WHAT YOU'RE DOING IT?!?!?!

I would very much like to do this in my preferred command line.

This definitely worked on bash on my machine! Holy moly!


[NOTE: I use JPSoft's TakeCommand {TCC.EXE} as my preferred CLI since the 1980s, though it was called something else back then - Norton Dos, so i have access to most internal/unix like commands but not necessarily internal powershell commands]

**Reply by j4james** [reply]
In case it helps, this is a gist with both commands in a simple text file:
https://gist.github.com/j4james/b22fa1237c724e74ff24b6fd3d63babd

If you download the raw content, you should just be able to `type` or `cat` that in any shell to see the results.

It's just plain ASCII with some escape sequences.
1. The first part is a [`DECDLD`](https://vt100.net/docs/vt510-rm/DECDLD) sequence which defines a custom character set (also known as a soft font) with four characters, `()*+`, representing the trumpet glyphs.
2. To activate that character set, you use an [`SCS`](https://vt100.net/docs/vt510-rm/SCS) sequence, with the identifier ` @`, i.e. `\x1B( @` (note that there's a space before the `@`)
3. Then you can output the left trumpet with `*(` and the right trumpet with `)+`.
4. But when you want to output regular text again, you need to switch back to the ASCII character sets with another `SCS` sequence, `\x1B(B`.

**Reply by ClaireCJS** [reply]
Can identifiers other than "@" be used?  I want to load 2 of these at once so I can access both 10x20 and 16x32. [Is that asking too much? 😅 Sometimes I obsess for perfection.]

**Reply by j4james** [reply]
You can use different identifiers, but that won't allow you to create two different fonts. We only support a single buffer, so as soon as you try and create the second font, it'll overwrite the first.

That said, the standard does allow for multiple buffers, so that's theoretically something we could support in the future.

**Comment by ClaireCJS** [comment]
I got it working!! Thanks!!

Got hung up for 15 minutes on the fact that TCC has it's own internal printf which does not behave the same as cygwin's more official printf.exe

**thank you so much!!!**

**Reply by j4james** [reply]
I'm glad you got working. But I replied above with some more details in case you're interested in how it works.

**Reply by ClaireCJS** [reply]
I'm absolutely interested enough that I'm going to be transcribing this into my personal notes! Thank you so much!!

**Reply by ClaireCJS** [reply]
Additional question - this method only works with the trumpet character - there's no way to simply "flip backwards on" so all characters are flipped, then "flip backwards off", is there?

So this would have to be extended per-character that I want flipped, right?

That's fine! I'm just curious.

**Reply by j4james** [reply]
It's not specific to the trumpet character at all. I've literally defined the glyphs myself, pixel by pixel, to look like the trumpet character. It could be anything though.

The limitation is that you can only defined 94 characters in a set, and we currently only support one set. If you run another application that also creates a custom character set, your characters will then get replaced (unless you recreate them every time you need them).

**Reply by ClaireCJS** [reply]
Wow that's amazing!  Console has come so far, but it seems a lot of console's progress is simply implementing all the pre-existing standards that were never done properly! Ha!

I'm very happy now.  The custom character sets can also be ansi colored UNLIKE the original emojis.

So you didn't just unlock flipped trumpets for me, you unlocked rainbow trumpets for me!  🤣🌈

Wow thank you! This is so cool! I used to make ANSI art back in THEDRAW.EXE days and it's so cool to see things suddenly progress so far, so fast.

![image](https://github.com/microsoft/terminal/assets/789591/dfdef5b9-9ecc-4cbf-9dcc-dc9666646692)

**Reply by ClaireCJS** [reply]
This stuff is amazing. I've been playing with this for a good hour now.

My next question is, how were you able to create that sequence?

Did you have to somehow decode the trumpet into a pixel representation?

Like, I want to do the same for EMOJI_DEVIL_FACE so that I can color it blue. But I'm not sure how. I'm not sure how to easily extend this idea to other already-existing emojis (hopefully without having to "Draw" them pixel by pixel).

I'm probably looking in the wrong place for answers, but i'm wondering if there is a Tutorial or Wizard for creating that sequence somehow.  I'd definitely like to do it for devil_face and castle and bat. [for dumb personal reasons]

**Reply by j4james** [reply]
The closest thing to a tutorial that I can think of is this site:
https://vt100.net/dec/vt320/soft_characters

It's based on the VT320 which has a weird 15x12 cell size, whereas I tend to use a simpler 10x20 cell. However, the basic concept is the same, and they've got a nice section near the end ("Glyph Definition"), which explains the character-to-pixel mapping quite well.

Once you've read that, you should be able to understand the first sixel sequence from my trumpet font, which generates half of the right facing trumpet glyph like this:

![image](https://github.com/microsoft/terminal/assets/4181424/d26fe8b3-e8b5-4fc2-9979-d55a2a942f0d)

I have a little editor that I use for this, but it's a big hack which wouldn't really be usable by anyone but me. One day I hope to turn it into something fit for human consumption, but that is probably a long way off.

Edit: I've been doing some research, and I found a tool that does font editing and conversion, and it claims to support the DEC soft font format. I haven't tried it myself, but it looks promising.
https://github.com/robhagemans/monobit

**Reply by ClaireCJS** [reply]
I totally want to try your editor. But I finally understand this! Thank you so very much!  Can't believe I never heard of 'sixels' before, it's a pretty awesome concept.

**Reply by ClaireCJS** [reply]
One more thing -- do you know how to do these in color? I'm seeing that it can be done with color per-pixel. I can't get it to really work though, maybe if i had a working example.

**Reply by j4james** [reply]
The original DEC terminals didn't support multicolor fonts, but there were some that supported color images using a similar sixel encoding. And I think there are one or two modern terminals that adapted that format for use in fonts as well. Windows Terminal doesn't support that, though, and I don't know how exactly it works.

But if you want to find out more, I'd suggest looking at MLTerm or RLogin. They're the most likely terminals to have multicolor fonts, and they must have some documentation somewhere on the subject.

**Reply by j4james** [reply]
@ClaireCJS If you're still interested in trying my font editor, I've just uploaded the repo to github here: [vtfontmaker](https://github.com/j4james/vtfontmaker). It still needs a bit of work, but it's mostly usable. The only problem is that it requires a terminal with paging functionality, and that has only recently been merged in Windows Terminal, so you'll have to have a recent [nightly build](https://github.com/microsoft/terminal?tab=readme-ov-file#installing-windows-terminal-canary).
